### PR TITLE
feat: reimplement hostpath local PV with CSI

### PR DESCRIPTION
**Pull request by zjx20**

### CODE REVIEW

File: pkg/csi/hostpathimpl.go [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+		if err != nil {
+			_ = impl.ns.osTool.Remove(volumeHostPath)
+		}
+	}()
+
+	// set project quota
+	setQuotaDone := false
+	var setQuotaErr error
+	tags := tags(req.VolumeContext)
+	capacity, err := tags.GetInt64(volumeCapacityTag)
```
Comment (gufeijun) [inline pkg/csi/hostpathimpl.go:133]: better to check if capacity is less than zero

File: pkg/csi/hostpathimpl.go [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+		}
+	}
+	enforceCapacityLimit, _ := tags.GetBool(enforceCapacityLimitScTag)
+	if enforceCapacityLimit && !setQuotaDone {
+		return nil, status.Errorf(codes.Internal,
+			"NodePublishVolume: failed to set project quota, err: %s", setQuotaErr)
+	}
+
+	// mount target path
+	if err := impl.mountHostPathVolume(ctx, req, volumeHostPath); err != nil {
```
Comment (gufeijun) [inline pkg/csi/hostpathimpl.go:151]: err is shadowed, so if err at line 134 is not nil, may cause the volume successfully mounted but then be removed in line 125. Or fail to mount, but the volume directory is not removed.

File: deploy/helm/templates/storage-class.yaml [hunk 3]
```diff
@@ -69,7 +69,7 @@ allowVolumeExpansion: true
 apiVersion: storage.k8s.io/v1
 kind: StorageClass
 metadata:
-  name: {{ .Values.storageclass.hostpath.name }}
+  name: {{ .Values.storageclass.hostpath.name }}-legacy
   annotations:
```
Comment (cnut) [inline deploy/helm/templates/storage-class.yaml:73]: is legacy storage class still useful?

File: pkg/csi/client/grpc.go [hunk 4]
```diff
@@ -42,7 +42,7 @@ type Connection interface {
 	CreateSnapshot(ctx context.Context, vgName string, snapshotName string, srcVolumeName string, readonly bool, roInitSize int64, secrets map[string]string) (int64, error)
 	DeleteSnapshot(ctx context.Context, volGroup string, snapVolumeID string, readonly bool, secrets map[string]string) error
 	ExpandVolume(ctx context.Context, volGroup string, volumeID string, size uint64) error
-	CleanPath(ctx context.Context, path string) error
```
Comment (cnut) [inline pkg/csi/client/grpc.go:45]: can we extend DeleteVolume to clean path?

File: pkg/csi/hostpathimpl.go [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+	if volumeSource := req.GetVolumeContentSource(); volumeSource != nil {
+		return nil, status.Error(codes.Unimplemented, "CreateVolume: volume content source is not supported")
+	}
+
+	reqCtx := getValue(ctx, ctxKeyCreateVolume)
+	nodeName := reqCtx.nodeName
+	volumeID := req.GetName()
+	parameters := req.GetParameters()
+
+	// TODO(x.zhou): interact with the scheduler to see if there is sufficient space to allocate
```
Comment (cnut) [inline pkg/csi/hostpathimpl.go:67]: if there is no sufficient space to allocate this volume, we can only return error, then the user will be stucked here, so we need extend schedule intead of check here?

Am I right?

File: pkg/csi/hostpathimpl.go [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+	if err != nil {
+		return fmt.Errorf("mountHostPathVolume: check if targetPath %s is mounted: %s", targetPath, err.Error())
+	}
+	if !notmounted {
+		log.Infof("mountHostPathVolume: volume %s(%s) is already mounted", req.VolumeId, targetPath)
+		return nil
+	}
+
+	// start to mount
+	mnt := req.VolumeCapability.GetMount()
```
Comment (cnut) [inline pkg/csi/hostpathimpl.go:185]: k8s hostpath example also check whether the target path is existing or not

File: pkg/csi/hostpathimpl.go (hunk 1 above)
Comment (zjx20) [reply pkg/csi/hostpathimpl.go:139]: Done.

**Review by cnut — APPROVED** [review]

File: pkg/csi/hostpathimpl.go [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+	hostPathScTag             = "hostPath"
+	enforceCapacityLimitScTag = "enforceCapacityLimit"
+
+	// internal tags, the prefix is added to avoid conflicts with other public tags
+	volumeCapacityTag = "_hp_volumeCapacity"
+)
+
+// hostPathImpl implements the CSI interfaces for hostPath volume type
+type hostPathImpl struct {
+	baseControllerServer
```
[the remark this answers is not in the document]
Comment (cnut) [reply pkg/csi/hostpathimpl.go:48]: good job

File: pkg/csi/hostpathimpl.go [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+	log "k8s.io/klog/v2"
+
+	"github.com/alibaba/open-local/pkg"
+)
+
+const (
+	hostPathBasePathEnv = "HOSTPATH_BASE_PATH"
+
+	// tags from StorageClass
+	hostPathScTag             = "hostPath"
```
Comment (gufeijun) [inline pkg/csi/hostpathimpl.go:39]: what about add a prefix like "csi.openlocal.kubeblocks.io/"

File: pkg/csi/hostpathimpl.go [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+	})
+	if err != nil {
+		err = fmt.Errorf("setProjectQuota failed, error: %w, output: %s", err, out)
+	}
+	log.Infof("setProjectQuota: cmd output: %s, err: %v", out, err)
+	return err
+}
+
+func getVerifiedHostPath(volumeID string, params map[string]string) (string, string, error) {
+	basePath := params[hostPathScTag]
```
Comment (gufeijun) [inline pkg/csi/hostpathimpl.go:261]: Maybe it's better to move validation check of arguments to CreateVolume stage? Expose error as quick as possible.

File: pkg/csi/hostpathimpl.go (hunk 8 above)
Comment (zjx20) [reply pkg/csi/hostpathimpl.go:39]: These fields are placed in `storageClass.parameters`, indicating that they are specific to this driver, so there is no need to add a prefix. If they are in annotations, it is necessary to add one.

File: pkg/csi/hostpathimpl.go (hunk 9 above)
Comment (zjx20) [reply pkg/csi/hostpathimpl.go:267]: It definitely should be, I missed it. Done.

File: pkg/csi/hostpathimpl.go (hunk 2 above)
Comment (gufeijun) [reply pkg/csi/hostpathimpl.go]: https://go.dev/play/p/LzFI_sdYmO8
image
if return err of "mountHostPathVolume" is not nil, may cause the clean job in defer can not be triggered

Comment (gufeijun) [reply pkg/csi/hostpathimpl.go]: image

Comment (zjx20) [reply pkg/csi/hostpathimpl.go]: Please be aware that the code you've provided is not identical to the code in the PR, which explicitly returns an error, so the correct err value can be seen within the defer statement.

image

**Pull request merged**